Add support for Dynamic LSTM quantization on Mobile

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32757 Add support for Dynamic LSTM quantization on Mobile**

Summary:
This PR updates the main quantize_dynamic API to use QNNPACK backend for mobile

Test Plan:
python test/test_quantization.py PostTrainingDynamicQuantTest.test_quantized_rnn

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D19632220](https://our.internmc.facebook.com/intern/diff/D19632220)

supriyar committed:
Add support for Dynamic LSTM quantization on Mobile

Summary:
This PR updates the main quantize_dynamic API to use QNNPACK backend for mobile

Test Plan:
python test/test_quantization.py PostTrainingDynamicQuantTest.test_quantized_rnn

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "Add support for Dynamic LSTM quantization on Mobile"


Summary:
This PR updates the main quantize_dynamic API to use QNNPACK backend for mobile

Test Plan:
python test/test_quantization.py PostTrainingDynamicQuantTest.test_quantized_rnn

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit b69fa375:


* **3/3** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=b69fa3759b4dc5a57fa3f0768d0e696f232422c7).

### :detective: 3 new failures recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4366621) caffe2_onnx_py2_gcc5_ubuntu16_04_build (1/3)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=92400796) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=92400796))
<details>
<summary>
<code>Jan 29 02:22:38 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:289:71: error: no matching function for call to 'cast(const at::Tensor&)'</code>
</summary>

```
Jan 29 02:22:38 /var/lib/jenkins/workspace/aten/src/ATen/cpp_custom_type_hack.h:23:4: note:   template argument deduction/substitution failed: 
Jan 29 02:22:38 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:287:67: error: template argument 1 is invalid 
Jan 29 02:22:38            cpp_custom_type_hack::cast<PackedLinearWeight>(params[i]); 
Jan 29 02:22:38                                                                    ^ 
Jan 29 02:22:38 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:289:38: error: the value of 'PackedLinearWeight' is not usable in a constant expression 
Jan 29 02:22:38            cpp_custom_type_hack::cast<PackedLinearWeight>(params[i + 1]); 
Jan 29 02:22:38                                       ^ 
Jan 29 02:22:38 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:287:38: note: 'PackedLinearWeight' was not declared 'constexpr' 
Jan 29 02:22:38            cpp_custom_type_hack::cast<PackedLinearWeight>(params[i]); 
Jan 29 02:22:38                                       ^ 
Jan 29 02:22:38 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:289:71: error: no matching function for call to 'cast(const at::Tensor&)' 
Jan 29 02:22:38            cpp_custom_type_hack::cast<PackedLinearWeight>(params[i + 1]); 
Jan 29 02:22:38                                                                        ^ 
Jan 29 02:22:38 In file included from /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:5:0: 
Jan 29 02:22:38 /var/lib/jenkins/workspace/aten/src/ATen/cpp_custom_type_hack.h:23:4: note: candidate: template<class T> T& at::cpp_custom_type_hack::cast(const at::Tensor&) 
Jan 29 02:22:38  T& cast(const Tensor& packed) { 
Jan 29 02:22:38     ^ 
Jan 29 02:22:38 /var/lib/jenkins/workspace/aten/src/ATen/cpp_custom_type_hack.h:23:4: note:   template argument deduction/substitution failed: 
Jan 29 02:22:38 In file included from /var/lib/jenkins/workspace/build/aten/src/ATen/core/TensorBody.h:5:0, 
Jan 29 02:22:38                  from /var/lib/jenkins/workspace/aten/src/ATen/Tensor.h:11, 
Jan 29 02:22:38                  from /var/lib/jenkins/workspace/aten/src/ATen/Context.h:4, 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4366604) pytorch_linux_xenial_py3_clang5_mobile_build (2/3)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=92400851) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=92400851))
<details>
<summary>
<code>Jan 29 02:22:43 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:289:71: error: no matching function for call to 'cast(const at::Tensor&)'</code>
</summary>

```
Jan 29 02:22:43 /var/lib/jenkins/workspace/aten/src/ATen/cpp_custom_type_hack.h:23:4: note:   template argument deduction/substitution failed: 
Jan 29 02:22:43 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:287:67: error: template argument 1 is invalid 
Jan 29 02:22:43            cpp_custom_type_hack::cast<PackedLinearWeight>(params[i]); 
Jan 29 02:22:43                                                                    ^ 
Jan 29 02:22:43 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:289:38: error: the value of 'PackedLinearWeight' is not usable in a constant expression 
Jan 29 02:22:43            cpp_custom_type_hack::cast<PackedLinearWeight>(params[i + 1]); 
Jan 29 02:22:43                                       ^ 
Jan 29 02:22:43 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:287:38: note: 'PackedLinearWeight' was not declared 'constexpr' 
Jan 29 02:22:43            cpp_custom_type_hack::cast<PackedLinearWeight>(params[i]); 
Jan 29 02:22:43                                       ^ 
Jan 29 02:22:43 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:289:71: error: no matching function for call to 'cast(const at::Tensor&)' 
Jan 29 02:22:43            cpp_custom_type_hack::cast<PackedLinearWeight>(params[i + 1]); 
Jan 29 02:22:43                                                                        ^ 
Jan 29 02:22:43 In file included from /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:5:0: 
Jan 29 02:22:43 /var/lib/jenkins/workspace/aten/src/ATen/cpp_custom_type_hack.h:23:4: note: candidate: template<class T> T& at::cpp_custom_type_hack::cast(const at::Tensor&) 
Jan 29 02:22:43  T& cast(const Tensor& packed) { 
Jan 29 02:22:43     ^ 
Jan 29 02:22:43 /var/lib/jenkins/workspace/aten/src/ATen/cpp_custom_type_hack.h:23:4: note:   template argument deduction/substitution failed: 
Jan 29 02:22:43 In file included from /var/lib/jenkins/workspace/build_mobile/aten/src/ATen/core/TensorBody.h:5:0, 
Jan 29 02:22:43                  from /var/lib/jenkins/workspace/aten/src/ATen/Tensor.h:11, 
Jan 29 02:22:43                  from /var/lib/jenkins/workspace/aten/src/ATen/Context.h:4, 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4366610) caffe2_onnx_py3_6_clang7_ubuntu16_04_build (3/3)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=92400867) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=92400867))
<details>
<summary>
<code>Jan 29 02:22:51 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:292:39: error: no member named 'value_or' in 'at::Tensor'</code>
</summary>

```
Jan 29 02:22:51 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:289:38: error: unknown type name 'PackedLinearWeight'; did you mean 'PackedLinearWeightsQnnp'? 
Jan 29 02:22:51           cpp_custom_type_hack::cast<PackedLinearWeight>(params[i + 1]); 
Jan 29 02:22:51                                      ^~~~~~~~~~~~~~~~~~ 
Jan 29 02:22:51                                      PackedLinearWeightsQnnp 
Jan 29 02:22:51 /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qnnpack_utils.h:22:8: note: 'PackedLinearWeightsQnnp' declared here 
Jan 29 02:22:51 struct PackedLinearWeightsQnnp { 
Jan 29 02:22:51        ^ 
Jan 29 02:22:51 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:291:39: error: no member named 'value_or' in 'at::Tensor' 
Jan 29 02:22:51       bias_ih = packed_struct_ih.bias.value_or(undefined); 
Jan 29 02:22:51                 ~~~~~~~~~~~~~~~~~~~~~ ^ 
Jan 29 02:22:51 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:292:39: error: no member named 'value_or' in 'at::Tensor' 
Jan 29 02:22:51       bias_hh = packed_struct_hh.bias.value_or(undefined); 
Jan 29 02:22:51                 ~~~~~~~~~~~~~~~~~~~~~ ^ 
Jan 29 02:22:51 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:1098:30: warning: 'get_contiguous_memory_format' is deprecated [-Wdeprecated-declarations] 
Jan 29 02:22:51     gog = at::zeros_like(cx, LEGACY_CONTIGUOUS_MEMORY_FORMAT); 
Jan 29 02:22:51                              ^ 
Jan 29 02:22:51 /var/lib/jenkins/workspace/c10/core/MemoryFormat.h:32:46: note: expanded from macro 'LEGACY_CONTIGUOUS_MEMORY_FORMAT' 
Jan 29 02:22:51 #define LEGACY_CONTIGUOUS_MEMORY_FORMAT c10::get_contiguous_memory_format() 
Jan 29 02:22:51                                              ^ 
Jan 29 02:22:51 /var/lib/jenkins/workspace/c10/core/MemoryFormat.h:34:1: note: 'get_contiguous_memory_format' has been explicitly marked deprecated here 
Jan 29 02:22:51 C10_DEPRECATED inline MemoryFormat get_contiguous_memory_format() { 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

supriyar committed:
Update on "Add support for Dynamic LSTM quantization on Mobile"


Summary:
This PR updates the main quantize_dynamic API to use QNNPACK backend for mobile

Test Plan:
python test/test_quantization.py PostTrainingDynamicQuantTest.test_quantized_rnn

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jianyuh commented on test/test_quantization.py:
```diff
@@ -535,200 +535,204 @@ def checkQuantized(model):
-        for out, ref_val in zip(final_hiddens_fp16, ref_hid):
-            torch.testing.assert_allclose(out, ref_val)
+            # Compare fp16 quantized to unquantized
+            output_fp16, final_hiddens_fp16 = cell_fp16(x, hiddens)
```
When `qengine` is `qnnpack`, can we pass the test for FP16 dynamic quantized LSTM? I don't think QNNPACK currently support FP16.

supriyar replied:
In that case if USE_FBGEMM is set to true then it will just execute fbgemm test. I don't think we call `gather_quantized_params_dynamic` for fp16 test, do we?

jianyuh replied:
For FP16, we will call `gather_quantized_params_fp16`:
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/RNN.cpp#L1226-L1231

supriyar replied:
@jianyuh then it should not depend on what qengine is set and follow the older code path.

jianyuh reviewed (commented):
Thanks for working on this!

supriyar commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -275,28 +275,59 @@ static std::vector<QuantizedCellParams> gather_quantized_params(TensorList param
+        auto& packed_struct_hh =
+            cpp_custom_type_hack::cast<PackedLinearWeightsQnnp>(params[i + 1]);
+
+        bias_ih = packed_struct_ih.bias;
```
Bias cannot be null for qnnpack. We set it to all zeros if it is passed in as null - https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qlinear_prepack.cpp#L144

jianyuh reviewed (approved):
LGTM! I am trying to land https://github.com/pytorch/pytorch/pull/32739. Let's land this PR after https://github.com/pytorch/pytorch/pull/32739.

PCerles commented:
I am seeing disparities between the results on iOS and Linux with the same frozen torchscript graph quantized with nnpack(1.5.0.dev20200212+cpu). The model is a stacked bidirectional LSTM model. I do not see any disparity with the unquantized model. Any ideas why this could be happening?

Quantization command:
```torch.backends.quantized.engine = 'qnnpack'
    model = torch.quantization.quantize_dynamic(
                model, {nn.LSTM}, dtype=torch.qint8
                )
```

supriyar commented:
We might have a bug in the ARM 64bit kernels which is currently being investigated.. cc @AshkanAliabadi

PCerles commented:
Any idea on timeline for resolution? Wondering whether I should retrain a smaller model or wait on fix for deployment.

AshkanAliabadi commented:
5 days.

On Thu, Feb 13, 2020, 10:26 AM PCerles <notifications@github.com> wrote:

> Any idea on timeline for resolution? Wondering whether I should retrain a
> smaller model or wait on fix for deployment.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/pytorch/pull/32757?email_source=notifications&email_token=AD24NKADJ7MNGRBQWXBNGILRCWGGLA5CNFSM4KM5PTR2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOELWCSOI#issuecomment-585902393>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD24NKHUZDZSSME2VKFASELRCWGGLANCNFSM4KM5PTRQ>
> .
>

PCerles commented:
@AshkanAliabadi thanks so much, saw it got merged in here (https://github.com/pytorch/pytorch/pull/33320)

AshkanAliabadi commented:
Hey no worries. Did it fix the issue you were having?

On Wed, Feb 19, 2020, 8:05 AM PCerles <notifications@github.com> wrote:

> @AshkanAliabadi <https://github.com/AshkanAliabadi> thanks so much, saw
> it got merged in here (#33320
> <https://github.com/pytorch/pytorch/pull/33320>)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/pytorch/pull/32757?email_source=notifications&email_token=AD24NKF7H6SFK7SKVF3EPDTRDVKENA5CNFSM4KM5PTR2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMILHBA#issuecomment-588297092>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD24NKBWSS7FSB6BUV5Q2DLRDVKENANCNFSM4KM5PTRQ>
> .
>

PCerles commented:
Not yet, now getting this issue:

```
2020-02-19 15:31:49.239037-0800 PyTorchDemo[10893:2599531] quantization scale should be > 0
The above operation failed in interpreter.
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/torch/nn/quantized/dynamic/modules/rnn.py", line 313

        if batch_sizes is None:
            result = _VF.quantized_lstm(input, hx, weight_values, self.bias, self.num_layers,
                     ~~~~~~~~~~~~~~~~~~ <--- HERE
                                        float(self.dropout), self.training, self.bidirectional,
                                        self.batch_first, dtype=self.dtype, use_dynamic=True)
Serialized   File "code/__torch__/torch/nn/quantized/dynamic/modules/rnn.py", line 81
      _28 = self.dtype
      _29, _30, = hx0
      _31, _32, _33 = torch.quantized_lstm(input, [_29, _30], weight_values, _22, _23, _24, _25, _26, _27, dtype=_28, use_dynamic=True)
                      ~~~~~~~~~~~~~~~~~~~~ <--- HERE
      result = (_31, _32, _33)
    else:
No valid output!
```
(again, only on ARM64 for iOS, not with qnnpack quantized model on linux), with commit `43e015f4b1c6f688ebf0f1bc573f51e53ecef896`

AshkanAliabadi commented:
@supriyar Do you know anything about this, or is this another bug in the kernels? :/

supriyar commented:
The problem seems to be here `PyTorchDemo[10893:2599531] quantization scale should be > 0`

@PCerles can you check the input values to see if they seem fine? The scale is computed dynamically so if possible try printing the input or `max`/`min` here to see if they make sense. https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/quant_utils.h#L60
 
cc @jianyuh as we saw something similar on FBGEMM as well.

supriyar commented:
If that doesn't work and you have a repro, we can take a look. Please file an issue with the repro instructions.

PCerles commented:
No luck with that unfortunately. I should add the model outputs are completely wrong too compared to the unquantized model (though again, only on arm64, not with qnnpack engine on linux). I will work up a repro

PCerles commented:
@supriyar can I send you a zip file with model / XCode Proj? Relatively big (unquantized model ~200Mb)

PCerles commented:
Hey @supriyar @AshkanAliabadi, turns out I was linking to the wrong install of PyTorch. The PR seems to have fixed it. Thanks so much for your help! Can finally deploy some fast edge end-to-end speech recognition 👍

AshkanAliabadi commented:
Awesome! Super cool.

On Thu, Feb 20, 2020, 11:06 AM PCerles <notifications@github.com> wrote:

> Hey @supriyar <https://github.com/supriyar> @AshkanAliabadi
> <https://github.com/AshkanAliabadi>, turns out I was linking to the wrong
> install of PyTorch. The PR seems to have fixed it. Thanks so much for your
> help! Can finally deploy some fast edge end-to-end speech recognition 👍
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/pytorch/pull/32757?email_source=notifications&email_token=AD24NKAEU4PYVR2L5TMBCUTRD3IC5A5CNFSM4KM5PTR2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMPVNCQ#issuecomment-589256330>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD24NKFFQVOD54MVCUOYJKTRD3IC5ANCNFSM4KM5PTRQ>
> .
>
